### ISSUE TITLE
Question about sleep button 

### ISSUE THREAD

**Issue by CaiHengrui** [CaiHengrui/BME590L-lab07#6]
https://github.com/CaiHengrui/BME590L-lab07/blob/44be7839019728cc6d3512d1473c03503e2dd607/src/main.c#L1-L167
– At any point in time, if the sleep button is pressed, all LEDs, except for the heartbeat LED,
go off until the sleep button in pressed again, at which time the device returns back to the exact state it was in before being put to sleep.

Don't know how to write codes about being back to previous state?

**Comment by CaiHengrui** [comment]
Can I use `GPIO_DT_SPEC_GET` in the loop to acquire previous status?

**Comment by CaiHengrui** [comment]
Can I use the function "gpio_pin_configure_dt " in the while loop

**Comment by mlp6** [comment]
Why do you want to "reconfigure" the pin in the while loop?

**Comment by mlp6** [comment]
As for the previous state, you need to be able to store what state the LEDs were in (on/off), and when resuming, what rate they should be blinking at.

**Comment by CaiHengrui** [comment]
No,I don't want to reconfigure. I hope I can capture the status(configuration?) What function should I use? or I don't need a function?

**Comment by mlp6** [comment]
`gpio_pin_configure_dt()` is a configuration function; hence why I asked that.  Since you can explicitly set the LED state, you can/should record what it is set to when you perform that operation.

**Comment by CaiHengrui** [comment]
So I need to analyze on my own what status they should be? 
btw, if the whole thing is too complicated to analyze, is there a way we can use a kind of function to acquire the status?

**Comment by CaiHengrui** [comment]
Sorry I'm confused. How can I explicitly set the LED state？

**Comment by CaiHengrui** [comment]
void set_ble_leds_rgb(bool red, bool green, bool blue);
void set_meas_leds_rgb(bool red, bool green, bool blue);
void set_battery_leds(int level);
like this？

**Comment by mlp6** [comment]
https://docs.zephyrproject.org/3.1.0/hardware/peripherals/gpio.html#c.gpio_pin_set_dt

**Comment by CaiHengrui** [comment]
I thought the `gpio_pin_set_dt` can set the LED state and update it? How can I get the previous state if I set some LED,I mean, go back to the status I have not set?
https://github.com/CaiHengrui/BME590L-lab07/blob/44be7839019728cc6d3512d1473c03503e2dd607/src/main.c#L155-L165
Can I return those values and that means I can acquire the previous state?

**Comment by CaiHengrui** [comment]
or can I use `gpio_pin_get` to read the current state of LEDs, then I use`gpio_pin_set_dt` to set it to sleep mode. Meanwile i still have the previous state to use.

**Comment by mlp6** [comment]
I think that `gpio_pin_get()` only works on input pins: https://docs.zephyrproject.org/3.1.0/hardware/peripherals/gpio.html#c.gpio_pin_get

**Comment by mlp6** [comment]
Again, if you are setting an output pin state explicitly, then you know its state and can save that in a variable.

**Comment by CaiHengrui** [comment]
But how can I know its stats if after I set those pins state ,those pins state changes somehow maybe because I pressed other buttons. So I do not know the present state, I mean, the code would't know because it's undefined which button I would presse.

 I just need to press the sleep button, and some codes will save the output pin state in that moment into a variable in that moment. After I press the sleep button again, I can use `gpio_pin_set_dt` to retrieve the saved pin state.

Therefore, the question is ,I can also use `gpio_pin_set_dt` to save the output pin state in that moment ?

**Comment by CaiHengrui** [comment]
![图片](https://user-images.githubusercontent.com/122759256/221703072-f21057ee-c8ce-4abc-85cc-ed01d58a8d81.png)
Is that correct?

**Comment by mlp6** [comment]
No - that function returns simply an exit code on success of setting or not.

But I still maintain that you know the state of all of the other LEDs because you are setting them... at some point in your code, you are setting pin states, and it remains in that state until you set it to something else, so you should be able to bookeep that.

**Comment by mlp6** [comment]
by your logic above, you know all of the LED states since you are setting them via your `val_*` variables

**Comment by CaiHengrui** [comment]
so I can not save the states into `val_` variables, I must need to know them at first.

**Comment by CaiHengrui** [comment]
Thank you so much. I will try agian.

**Comment by mlp6** [comment]
correct, *you* control the state of the LEDs
